oneflow/core/eager/eager_blob_object.cpp

```diff
@@ -83,5 +87,326 @@ Maybe<void> EagerBlobObject::TryAllocateBlobBodyMemory(DeviceCtx* device_ctx) {
+        if (!dtr_blob_object->is_in_memory()) {
+          max = std::max(max, dtr_blob_object->child_depth());
+        }
+      }
+    }
+  }
+  return max + 1;
+}
+
+Maybe<double> DTREagerBlobObject::child_cost(bool is_bp_required) const {
+
+  double cost = 0;
+
+  for (int i = 0; i < user_ops_.size(); ++i) {
+    const auto* ptr = dynamic_cast<DTRInstrOperand*>(JUST(user_op(i)));
+    CHECK_NOTNULL_OR_RETURN(ptr);
+    for (const auto& output : ptr->outputs()) {
+      if (!output.expired()) {
+        auto object = output.lock();
+        const auto dtr_blob_object = std::dynamic_pointer_cast<vm::DTREagerBlobObject>(object);
+        CHECK_NOTNULL_OR_RETURN(dtr_blob_object);
+        bool add_flag = (!dtr_blob_object->is_in_memory());
+        if (is_bp_required) {
+          add_flag = add_flag && dtr_blob_object->is_bp_required();
+        }
+        if (add_flag) {
+          auto com_time = dtr_blob_object->compute_time();
+          auto c_cost = JUST(dtr_blob_object->child_cost(is_bp_required));
+          cost = cost + com_time + c_cost;
+        }
+      }
+
+      // CHECK_OR_RETURN(static_cast<bool>(output.get()));
+      // auto object = output.get();
+      // CHECK_NOTNULL_OR_RETURN(object);
+      // const auto* dtr_blob_object = dynamic_cast<vm::DTREagerBlobObject*>(output.get());
+      // // CHECK_NOTNULL_OR_RETURN(dtr_blob_object);
+      // if (dtr_blob_object == nullptr) {
+      //   continue;
+      // } else {
+      //   if (!dtr_blob_object->is_in_memory()) {
+      //     auto com_time = dtr_blob_object->compute_time();
+      //     auto c_cost = JUST(dtr_blob_object->child_cost());
+      //     cost = cost + com_time + c_cost;
+      //   }
+      // }
+    }
+  }
+
+  return cost;
+}
+
+Maybe<double> DTREagerBlobObject::neighbor_cost() const {
+  auto p_cost = JUST(parent_cost());
+  auto c_cost = JUST(child_cost());
+  return p_cost + c_cost + compute_time_;
+}
+
+Maybe<double> DTREagerBlobObject::approx_neighbor_cost() const {
+  double cost = 0;
+  const auto& inputs = compute_op_->inputs();
+  for (int i = 0; i < inputs.size(); ++i) {
+    if (auto tmp = inputs[i].lock()) {
+      auto dtr_blob_object = dynamic_cast<vm::DTREagerBlobObject*>(tmp.get());
+      CHECK_NOTNULL_OR_RETURN(dtr_blob_object);
+      if (!dtr_blob_object->is_in_memory()) {
+        double p_cost =
+            Global<one::DTRTensorPool>::Get()->find_father(dtr_blob_object->node)->compute_time();
+            if (p_cost < dtr_blob_object->compute_time()) {
+              p_cost = dtr_blob_object->compute_time();
+            }
+        cost += p_cost;
+      }
+    }
+  }
+
+  const auto& outputs = compute_op_->outputs();
+  for (int i = 0; i < outputs.size(); ++i) {
+    if (auto tmp = outputs[i].lock()) {
+      auto dtr_blob_object = dynamic_cast<vm::DTREagerBlobObject*>(tmp.get());
+      CHECK_NOTNULL_OR_RETURN(dtr_blob_object);
+      if (!dtr_blob_object->is_in_memory()) {
+        double c_cost =
```

Comment:
suppose
A -> B, C -> D, with A evicted, and we are trying to calculate the cost for D
if I get it correctly, it will be cost of B + cost of C, which add A twice?
A way to avoid this is to create a temporary node in union find.